python/ray/data/dataset.py

```diff
@@ -101,20 +101,55 @@
 
 @PublicAPI
 class Dataset(Generic[T]):
-    """Implements a distributed Arrow dataset.
-
-    Datasets are implemented as a list of ``ObjectRef[Block]``. The block
-    also determines the unit of parallelism. The default block type is the
-    ``pyarrow.Table``. Arrow-incompatible objects are held in ``list`` blocks.
+    """A Dataset is a distributed data collection for data loading and processing.
+
+    Datasets are implemented as a list of `ObjectRef[Block]`, where each block
+    holds an ordered collection of items, representing a shard of the overall
+    data collection. The block can be either a `pyarrow.Table`, or Python list.
+    The block also determines the unit of parallelism.
+
+    Datasets can be created in multiple ways: from synthetic data via `range_.*()`
+    APIs, from existing memory data via `from_*()` APIs, or from external storage
+    systems such as local disk, S3, HDFS etc. via the `read_*()` APIs. The
+    (potentially processed) Dataset can be saved back to external storage systems via
+    the `write_*()` APIs.
+
+    Examples:
+        >>> import ray
+        >>> # Create dataset from synthetic data.
+        >>> ds = ray.data.range(1000) # doctest: +SKIP
+        >>> # Create dataset from in-memory data.
+        >>> ds = ray.data.from_items( # doctest: +SKIP
+        ...     [{"col1": i, "col2": i * 2} for i in range(1000)])
+        >>> # Create dataset from external storage system.
+        >>> ds = ray.data.read_parquet("s3://bucket/path") # doctest: +SKIP
+        >>> # Save dataset back to external storage system.
+        >>> ds.write_csv("s3//bucket/output") # doctest: +SKIP
+
+    Datasets supports parallel processing at scale: transformations such as
+    `.map_batches()`, aggregations such as `.min()`/`.max()`/`.mean()`, grouping via
+    `.groupby()`, shuffling operations such as `.sort()`, `.random_shuffle()`, and
+    `.repartition()`.
```

Comment:
We haven't set `:py:obj` to be the default role for Sphinx, so these will need to be e.g. :meth:`.sort()`.